File: Java/3.java
/**
 * @no 3
 * @name Longest Substring without Repeating Characters
 */
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }

        int ans = 0, p = 0;
        Map<Character, Integer> map = new HashMap<Character, Integer>();

        for (int q = 0; q < s.length(); q++) {
            p = Math.max(p, map.keySet().contains(s.charAt(q)) ? (int) map.get(s.charAt(q)) + 1: 0);
            ans = Math.max(ans, q - p + 1);
            map.put(s.charAt(q), q);
        }

        return ans;
    }
}

File: Python/3.py
"""
@no 3
@name Longest Substring without Repeating Characters
"""
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s or len(s) == 0:
            return 0
        dic = {}
        left = 0
        max_length = 0
        for right in range(0, len(s)):
            left = max(left, dic[s[right]] + 1 if s[right] in dic.keys() else 0)
            max_length = max(max_length, right - left + 1)
            dic[s[right]] = right
        return max_length
